clients/cobol-lsp-vscode-extension/src/services/copybook/CopybookDownloadService.ts

```diff
@@ -560,4 +581,134 @@ export class CopybookDownloadService implements vscode.Disposable {
       }
     }
   }
+
+  public async downloadCopybookE4E(copybookProfile: CopybookProfile) {
+    const endevorExplorerApi = await E4ECopybookService.getE4EClient(
+      copybookProfile.documentUri,
+      this.outputChannel,
+    );
+    if (!endevorExplorerApi) return;
+
+    await this.handleMembers(endevorExplorerApi, copybookProfile);
+    await this.handleElements(endevorExplorerApi, copybookProfile);
+  }
+
+  public async handleMembers(
+    endevorExplorerApi: e4eResponse,
+    copybookProfile: CopybookProfile,
+  ): Promise<void> {
+    endevorExplorerApi.members.forEach((member) => {
+      if (member.member === copybookProfile.getCopybook()) {
+        try {
+          this.downloadDatasetE4E(
+            member.dataset,
+            endevorExplorerApi,
+            copybookProfile,
+          );
+        } catch (err: any) {
+          if (
+            CopybookDownloadService.needsUserNotification([copybookProfile])
+          ) {
+            vscode.window.showErrorMessage(err.message);
+          }
+        }
+      }
+    });
+  }
+
+  public async handleElements(
+    endevorExplorerApi: e4eResponse,
+    copybookProfile: CopybookProfile,
+  ): Promise<void> {
+    endevorExplorerApi.elements.forEach((element) => {
+      if (element.element === copybookProfile.getCopybook()) {
+        try {
+          this.downloadElementE4E(endevorExplorerApi, copybookProfile, element);
+        } catch (err: any) {
+          if (
+            CopybookDownloadService.needsUserNotification([copybookProfile])
+          ) {
+            vscode.window.showErrorMessage(err.message);
+          }
+        }
+      }
+    });
+  }
+
+  public async downloadElementE4E(
+    endevorApi: e4eResponse,
+    copybookProfile: CopybookProfile,
+    element: EndevorElement,
+  ) {
+    const pgm = getProgramNameFromUri(copybookProfile.documentUri);
```

Comment:
can this be safely used for non-file schemes?

I don't quite understand the use or program name. It's not unique identifier.

Reply:
tested for URL and file schemes no issues. Do you see any pitfall ?